File: Ungdung/Intro Sort/IntroSort.java
package Sort;
public class IntroSort {
    int[] a;
    int n;
    public IntroSort(int size) {
        this.a = new int[size];
        this.n = 0;
    }
    void push(int data) {
        if (n < a.length) {
            a[n++] = data;
            System.out.println("Added: " + data);
        } else {
            System.out.println("Array is full!");
        }
    }
    void swap(int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }
    void maxHeap(int i, int heapN, int low) {
        int temp = a[low + i - 1];
        int c;
        while (i <= heapN / 2) {
            c = 2 * i;
            if (c < heapN && a[low + c - 1] < a[low + c]) {
                c++;
            }
            if (temp >= a[low + c - 1]) {
                break;
            }
            a[low + i - 1] = a[low + c - 1];
            i = c;
        }
        a[low + i - 1] = temp;
    }
    void heapify(int low, int high, int heapN) {
        for (int i = heapN / 2; i >= 1; i--) {
            maxHeap(i, high, low);
        }
    }

    // Heapsort
    void heapsort(int low, int high) {
        int heapN = high - low;
        heapify(low, high, heapN);
        for (int i = heapN; i >= 1; i--) {
            swap(low, low + i);
            maxHeap(i, high, low);
        }
    }

    // insertsort
    void insertion(int left, int right) {
        for (int i = left + 1; i <= right; i++) {
            int key = a[i];
            int j = i - 1;
            while (j >= left && a[j] > key) {
                a[j + 1] = a[j];
                j--;
            }
            a[j + 1] = key;
        }
    }
    // tim phan tu median
    int find(int a1, int b1, int c1) {
        int max = Math.max(Math.max(a[a1], a[b1]), a[c1]);
        int min = Math.min(Math.min(a[a1], a[b1]), a[c1]);
        int median = max ^ min ^ a[a1] ^ a[b1] ^ a[c1];
        if (median == a[a1]) {
            return a1;
        }
        if (median == a[b1]) {
            return b1;
        }
        return c1;
    }
    // partition cua quicksort
    int partition(int low, int high) {
        int pivot = a[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (a[j] <= pivot) {
                i++;
                swap(i, j);
            }
        }
        swap(i + 1, high);
        return i + 1;
    }
    // sort
    void sortdatautil(int low, int high, int depthlimit) {
        if (high - low > 16) {
            if (depthlimit == 0) {
                heapsort(low, high);
                return;
            }
            depthlimit = depthlimit - 1;
            int pivot = find(low, low + ((high - low) / 2) + 1, high);
            swap(pivot, high);
            int p = partition(low, high);
            sortdatautil(low, p - 1, depthlimit);
            sortdatautil(p + 1, high, depthlimit);
        } else {
            insertion(low, high);
        }
    }

    // sort
    void sortdata() {
        int depthlimit = (int) (2 * Math.floor(Math.log(n) / Math.log(2)));
        sortdatautil(0, n - 1, depthlimit);
    }

    // in mang
    void display() {
        for (int i = 0; i < n; i++) {
            System.out.print(a[i] + " ");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        int[] inp = {2, 10, 24, 12, 85, 72, 91, 43};
        IntroSort sorter = new IntroSort(inp.length);
        for (int num : inp) {
            sorter.push(num);
        }
        sorter.sortdata();
        sorter.display();
    }
}

File: Ungdung/Intro Sort/IntroSort.py
import math
from heapq import heappush, heappop

class IntroSort:
    def __init__(self):
        self.a = []
    def heapsort(self):
        h = []
        for value in self.a:
            heappush(h, value)
        self.a = [heappop(h) for _ in range(len(h))]
    def insertion_sort(self, low, high):
        for i in range(low + 1, high + 1):
            key = self.a[i]
            j = i - 1
            while j >= low and self.a[j] > key:
                self.a[j + 1] = self.a[j]
                j -= 1
            self.a[j + 1] = key
    def partition(self, low, high):
        pivot = self.a[high]
        i = low - 1
        for j in range(low, high):
            if self.a[j] <= pivot:
                i += 1
                self.a[i], self.a[j] = self.a[j], self.a[i]
        self.a[i + 1], self.a[high] = self.a[high], self.a[i + 1]  
        return i + 1
    def median_of_three(self, a, b, c):
        A = self.a[a]
        B = self.a[b]
        C = self.a[c]
        if A <= B <= C or C <= B <= A:
            return b 
        if B <= A <= C or C <= A <= B:
            return a 
        return c
    def introsort(self, low, high, depth_limit):
        size = high - low 
        if size < 16:
            self.insertion_sort(low, high)
        elif depth_limit == 0:
            self.heapsort()
        else:
            pivot = self.median_of_three(low, low + size // 2, high)
            self.a[pivot], self.a[high] = self.a[high], self.a[pivot]
            partition_point = self.partition(low, high)
            self.introsort(low, partition_point - 1, depth_limit - 1)
            self.introsort(partition_point + 1, high, depth_limit - 1)
    def introsorts(self, low, high):
        depth_limit = 2 * math.floor(math.log2(high - low))
        self.introsort(low, high, depth_limit)

    def __str__(self):
        return str(self.a)
if __name__ == '__main__':
    sorter = IntroSort()
    sorter.a = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]  # Dữ liệu mẫu
    print("Before sorting:", sorter)
    sorter.introsorts(0, len(sorter.a) - 1)
    print("After sorting:", sorter)
